Optimize processing speed for owned and membership requests and updated user count retrieval

Mynster9361 opened the pull request:
Improve the user count retrieval by utilizing a built-in function, significantly reducing processing time from over a minute to just five seconds for 100 users. Added a bit of error handling (might want more and also implement retry count?)

Please test it and let me know what you think i have only tested it against 999 users and seems like that worked as intended but let me know what you think!

Mynster9361 commented:
@jflieben will take a look again a bit later and look into retro and some delay when/if getting throttled

For the throttling you have seen/experienced i believe you would run into the same issue with your current implementation of job execution if you increase the amount of threads used further.

Thing is it is much slower to run individual rest request which properly gives you enough wait time for microsofts throttle limits as seen here:

https://learn.microsoft.com/en-us/graph/throttling-limits

And i would not say it is easy to calculate the exact cost for an application based on how much it is requesting but it is possible, just hard😅

jflieben commented:
@Mynster9361 very true, but the new-graphQuery function includes handling for throttling. I wish MS had implemented 429's consistently, but using retries and longer backoffs each time has at least gotten this to work very consistently for 1 thread per 250 users/tenant (so far). I have only tested tenants until almost 3k, but am in the process of onboarding a 30k user tenant to see how many threads we can do there.

Mynster9361 commented:
@jflieben 
Implemented some extra error handling and retry for the batch request still need a few cleanup tasks on it but it is working as intended now.
Tested on a tenant with about 50K users unfortunately it seems like the report generation is not to happy with so many entries in the report generation also the sorting step for the report generation takes 15 min by it self

I have added some messages from the M365PermissionsVerbose.log in the commit message here:
https://github.com/jflieben/M365Permissions/pull/28/commits/f8d8420124104a0ceecd576637539a694606ef07

Also so you could see some timestamps for how long/short the report takes to run with this change, Takes about 20-25 min to get the data for the report

jflieben commented:
@Mynster9361 I've just completed the review, and implemented most of your new logic!

I removed or modified some of it;

- toplevel param (wasn't used)
- batching for PIM stuff
- optionality of the batch function
- the retry function (implemented this directly in the existing graph function)
- delay handling in the retry

so basically just very minor changes! Tested it and it was indeed faster, and also has the same result as my previous run so it didn't break anything :)

I didn't know how to modify your PR directly, so I did this in 2 main commits (and some minor bugfix ones after):

1. https://github.com/jflieben/M365Permissions/pull/30/commits/9be7698c12fafe1c7e7f171b8035d7d00bc5a35f
2. https://github.com/jflieben/M365Permissions/pull/30/commits/09069ec0848fd3f93aee8343d8bf886cc467fb0c

thank you for this very elegant style batch handling addition!
